lib/rubocop/cop/style/empty_line_after_guard_clause.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop enforces empty line after guard clause
+      #
+      # @example
+      #
+      #   # bad
+      #   def foo
+      #     return if need_return?
+      #     bar
+      #   end
+      #
+      #   # good
+      #   def foo
+      #     return if need_return?
+      #
+      #     bar
+      #   end
+      #
+      #   # good
+      #   def foo
+      #     if something?
+      #       do_something
+      #       return if need_return?
+      #     end
+      #   end
+      class EmptyLineAfterGuardClause < Cop
+        include RangeHelp
+
+        MSG = 'Add empty line after guard clause.'.freeze
+
+        def on_if(node)
+          return unless contains_guard_clause?(node)
+
+          if node.parent
+            return if node.parent.single_line?
+            return if next_sibling_guard_clause?(node)
+          end
+
+          return if next_line_empty?(node)
+
+          add_offense(node)
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            node_range = range_by_whole_lines(node.source_range)
+            corrector.insert_after(node_range, "\n")
+          end
+        end
+
+        private
+
+        def next_sibling_guard_clause?(node)
+          next_sibling = node.parent.children[node.sibling_index + 1]
+          return false if next_sibling.nil?
+
+          next_sibling.if_type? && contains_guard_clause?(next_sibling)
+        end
+
+        def next_line_empty?(node)
+          next_line = processed_source[node.last_line]
+
+          line_end?(next_line) || next_line.blank?
+        end
+
+        def line_end?(line)
+          line =~ /^\s*end/
```

Comment:
I think regular expression is not good solution in RuboCop. We should use S-expression instead.
For example, the regexp matches `end_something` method incorrectly. Probably we can fix the problem with a regexp (e.g. `/^\s*end\b/`), but probably the regexp also has some problems. I think Ruby's syntax is too complex to parse with regular expression.

Maybe the cop can check sibling index instead of the regular expression. I guess when sibling index is the last, the node does not need an empty line on the next line.
For example:

```
if foo
  return if bar # index: 1 , it need an empty line.
  foo # index: 2
  return if baz  # index: 3 (last), It does not need an empty line.
end
```

Reply:
@pocke thanks, good point!

However, I think that your solution have an issue - node can have no parent. In this case, cop will not work.

So, maybe I will fix regex instead? I think something like `\s*end(\b|\.)` [will work](http://rubular.com/r/XICnYLyUim)

Btw, I found this regex in existing cop. Should I fix it here or send separate pr?

Reply:
> However, I think that your solution have an issue - node can have no parent. In this case, cop will not work.



A node without a parent is always a root node. And a root node does not have a next node.

```ruby
# The `if` node is a root node, and it does not have a next node. 
return if bar
```

So, I guess the cop can simply ignore a node without parent.




> Btw, I found this regex in existing cop. Should I fix it here or send separate pr?
 

Good catch! Can you separate PR?

Reply:
@pocke fixed

> Good catch! Can you separate PR?
Sure. Will do it right after we will end with this request.